### ISSUE TITLE
[@typescript-eslint/no-shadow] false positives with static methods

### ISSUE THREAD

**Issue by dinofx** [typescript-eslint/typescript-eslint#2592]
False positive reported for static methods with a type parameter with the same name as the containing class' type parameter.

```TS
export class Wrapper<Wrapped> {
  private constructor(private readonly wrapped: Wrapped) {}

  unwrap(): Wrapped {
    return this.wrapped;
  }

  static create<Wrapped>(wrapped: Wrapped) {
    return new Wrapper<Wrapped>(wrapped);
  }
}
```

In a static method, the type parameters for the class are not in scope because there's no instance.

**Comment by bradzacher** [comment]
Yeah this is a difficult one.
The scope analysis tooling assumes that each scope follows a tree-like structure. In this instance here the static scope is a child of the class scope.

To make this work "properly", we'd have to restructure the class analysis to handle this. It'd be super weird to setup.

It's probably better to just handle this in the rule - check the shadow and ignore it if the method is `static`.

**Comment by dinofx** [comment]
It looks like the problem is `nestFunctionScope` is always setting the upper scope to `this.currentScope`.  If you're in a static method, the upper scope should be the class' parent scope, not the class, I think?

Conditional types had the same problem.  You had to visit some of the children under one scope, and the "else" child under the original scope.  Maybe it would be easier to have the upper score be a parameter instead of something that is pushed and popped.

**Comment by bradzacher** [comment]
The class creates its own scope to contain both the generic type parameter decls, and for the class name (also the class heritage is referenced from within this scope).

If you reference the class name from inside of the class, it references a variable defined inside of the class scope, not the variable defined in the parent scope.

If we were to use the parent scope of the class scope, that would break that relationship, as the static method would reference the class variable defined in the parent scope.

The true fix for this would be to nest an "instance" scope within the class scope, and then define the class generics and instance members within that child scope, and then define static members in the class scope itself.

However this would be a large deviation from the core `eslint-scope` algorithm, and it's hard to say what the ramifications would be. Rules make assumptions based on what `eslint-scope` does, so we unfortunately have to adhere to it.

**Comment by Ionaru** [comment]
Here's another example of what I believe is the same issue, it happens with any enum I declare, either in the global scope or inside a function.

**Global**
code:
```ts
enum SomeEnum {}
```
error:
```json
[
  {
    "column": 6,
    "endColumn": 14,
    "endLine": 1,
    "line": 1,
    "message": "'SomeEnum' is already declared in the upper scope.",
    "messageId": "noShadow",
    "nodeType": "Identifier",
    "ruleId": "no-shadow",
    "severity": 2
  }
]
```

**In a function**
code:
```ts
(() => {
    enum SomeEnum {}
})();
```
error:
```json
[
  {
    "column": 10,
    "endColumn": 18,
    "endLine": 2,
    "line": 2,
    "message": "'SomeEnum' is already declared in the upper scope.",
    "messageId": "noShadow",
    "nodeType": "Identifier",
    "ruleId": "no-shadow",
    "severity": 2
  }
]
```

**Comment by bradzacher** [comment]
Nope. You are just using the wrong rule.

https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code

There are many issues that mention this, as well as the v4 release notes.

**Comment by OliverJAsh** [comment]
Does this false positive have the same root cause or should I file a new bug?

```ts
import * as Photo from '..';
enum Page {
  // 'Photo' is already declared in the upper scope.eslint@typescript-eslint/no-shadow)
  Photo = 'photo',
}
```

**Comment by bradzacher** [comment]
That is not a false-positive. It's correct analysis.
https://github.com/typescript-eslint/typescript-eslint/issues/2721#issuecomment-718804636
